hail/src/main/scala/is/hail/expr/ir/EmitFunctionBuilder.scala

```diff
@@ -342,29 +342,28 @@ class EmitFunctionBuilder[F >: Null](
       "FunctionBuilder emission should happen on master, but happened on worker")
 
     new ((Int) => F) with java.io.Serializable {
-      @transient
-      @volatile private var f: F = null
+      @transient @volatile private var theClass: Class[_] = null
 
       def apply(idx: Int): F = {
         try {
-          if (f == null) {
+          if (theClass == null) {
             this.synchronized {
-              if (f == null) {
+              if (theClass == null) {
                 childClasses.foreach { case (fn, b) => loadClass(fn, b) }
-                f = loadClass(n, bytes).newInstance().asInstanceOf[F]
-                if (localHConf != null)
-                  f.asInstanceOf[FunctionWithHadoopConfiguration].addHadoopConfiguration(localHConf)
+                theClass = loadClass(n, bytes)
               }
             }
           }
+          val f = theClass.newInstance().asInstanceOf[F]
```

Comment:
@catoverdrive Can you double check this is right?  Problem I noticed: we were creating one instance of `theClass` but calling `setPartitionIndex` on it for each partition requesting it.  So if two partitions on the same node immediately called `apply(idx: Int)`, the second one would overwrite the initialization of the first, which is incorrect.  What's more, it appears they'd both be working against the same random state, which also seems wrong.  How did it ever work?  I caught this bug by adding  a code cache that increased the amount of sharing of `f`, exposing the bug.

Reply:
:-|